Sources/Crypto/BCrypt/BCryptDigest.swift

```diff
@@ -1,60 +1,108 @@
-        )
-        let digest = try BCryptAlgorithm(config: config).digest(message: plaintext.convertToData())
-        let serializer = BCryptSerializer(config: config, digest: digest)
-        return serializer.serialize()
+    enum Algorithm: String, RawRepresentable {
+        /// older version
+        case _2a = "$2a$"
+        /// format specific to the crypt_blowfish BCrypt implementation, identical to `2b` in all but name.
+        case _2y = "$2y$"
+        /// latest revision of the official BCrypt algorithm, current default
+        case _2b = "$2b$"
+
+        /// Salt's length
+        var saltCount: Int {
+            return 29
+        }
+
+        /// Checksum's length
+        var checksumCount: Int {
+            return 31
+        }
     }
 
-    /// Verifies an existing BCrypt hash matches the supplied plaintext value. Verification works by parsing the salt and version from
-    /// the existing digest and using that information to hash the plaintext data. If hash digests match, this method returns `true`.
+    /// Generates string (29 chars total) containing the algorithm information + the cost + base-64 encoded 22 character salt
     ///
-    ///     let hash = try BCrypt.hash("vapor", cost: 4)
-    ///     try BCrypt.verify("vapor", created: hash) // true
-    ///     try BCrypt.verify("foo", created: hash) // false
+    ///     E.g:  $2b$05$J/dtt5ybYUTCJ/dtt5ybYO
+    ///           $AA$ => Algorithm
+    ///              $CC$ => Cost
+    ///                  SSSSSSSSSSSSSSSSSSSSSS => Salt
     ///
-    /// - parameters:
-    ///     - plaintext: Plaintext data to digest and verify.
-    ///     - hash: Existing BCrypt hash to parse version, salt, and existing digest from.
-    /// - throws: `CryptoError` if hashing fails or if data conversion fails.
-    /// - returns: `true` if the hash was created from the supplied plaintext data.
-    public func verify(_ plaintext: LosslessDataConvertible, created hash: LosslessDataConvertible) throws -> Bool {
-        let parser = try BCryptParser(serialized: hash.convertToData())
-        let digest = try BCryptAlgorithm(config: parser.parseConfig()).digest(message: plaintext.convertToData())
-        return try digest == parser.parseDigest()
+    /// Allowed charset for the salt: [./A-Za-z0-9]
+    private func generateSalt(cost: UInt, algorithm: Algorithm = ._2y) throws -> String {
+        let random = try URandom().generateString(count: 16)
+
+        guard let saltRaw = crypt_gensalt(
+            algorithm.rawValue,
+            cost,
+            random,
+            Int32(random.count) //Int32(entropy.utf8.count / MemoryLayout<UInt8>.size)
+        ) else {
+            throw CryptoError(identifier: "unableToGenerateSalt", reason: "Unable to generate BCrypt salt")
+        }
+
+        return String(cString: saltRaw)
+    }
+
+    public func hash(_ message: String, cost: UInt = 12) throws -> String {
```

Comment:
All public methods must have doc blocks that explain all parameters and have code examples.

Reply:
Yeah, ignore this PR. Submitting the OpenBSD version